<launch>

  <!-- Common frame/topic args for easy switching if TF frames differ -->
  <arg name="map_frame" default="map"/>
  <arg name="robot_frame" default="sensor"/>
  <arg name="robot_pose_topic" default="/state_estimation"/>

  <node pkg="dummy_vlm" type="dummyVLM" name="dummyVLM" output="screen" required="true">
    <param name="waypoint_file_dir" type="string" value="$(find dummy_vlm)/data/waypoints.ply" />
    <param name="object_list_file_dir" type="string" value="$(find dummy_vlm)/data/object_list.txt" />
    <param name="waypointReachDis" type="double" value="1.0" />
    <param name="map_frame" value="$(arg map_frame)"/>
    <param name="robot_frame" value="$(arg robot_frame)"/>
    <!-- Stagnation detection and recovery tuning -->
    <param name="stagnation_time" value="20.0"/>
    <param name="min_goal_delta" value="0.20"/>
    <param name="min_movement" value="0.20"/>
    <param name="recovery_cooldown" value="10.0"/>
    <param name="turn_radians" value="3.2"/>
    <param name="recovery_hold_sec" value="3.0"/>
    <!-- Recovery sequence tuning: back up then turn -->
    <param name="recovery_backup_dist" value="1"/>
    <param name="recovery_step_wait_sec" value="1.5"/>
    <param name="recovery_turn_wait_sec" value="1.0"/>
    <param name="recovery_reach_thresh" value="0.15"/>
  </node>
  <!-- Simple converter: /terrain_map (PointCloud2) -> /map (OccupancyGrid) -->
  <node pkg="dummy_vlm" type="pc_to_grid" name="pc_to_grid" output="screen">
    <param name="input_cloud" value="/terrain_map"/>
    <param name="map_frame" value="$(arg map_frame)"/>
    <param name="resolution" value="0.2"/>
    <param name="size_x" value="80.0"/>
    <param name="size_y" value="80.0"/>
    <param name="z_min" value="-1.0"/>
    <param name="z_max" value="2.0"/>
    <param name="occ_height_threshold" value="0.25"/>
    <param name="free_height_threshold" value="0.15"/>
    <param name="downsample" value="0.1"/>
  </node>
  <node pkg="dummy_vlm" type="frontier_explorer" name="frontier_explorer" output="screen">
    <param name="map_topic" value="/map"/>
    <param name="goal_topic" value="/next_goal"/>
    <param name="map_frame" value="$(arg map_frame)"/>
    <param name="robot_frame" value="$(arg robot_frame)"/>
    <param name="robot_pose_topic" value="$(arg robot_pose_topic)"/>
    <param name="fallback_enabled" value="true"/>
    <param name="fallback_radius_max" value="5.0"/>
    <param name="robot_radius" value="0.35"/>
    <param name="safety_margin" value="0.25"/>
    <!-- Frontier tuning -->
    <param name="min_frontier_cells" value="12"/>
    <param name="free_threshold" value="20"/>
    <param name="occ_threshold" value="65"/>
    <!-- Goal selection & hysteresis -->
    <param name="min_goal_dist" value="1.0"/>
    <param name="min_publish_separation" value="0.6"/>
    <param name="min_publish_period" value="1.5"/>
    <param name="unknown_radius" value="2.0"/>
    <param name="weight_unknown" value="1.0"/>
    <param name="weight_distance" value="0.2"/>
    <!-- Anti-oscillation / commitment tuning -->
    <param name="switch_improvement" value="5.0"/>
    <param name="goal_timeout" value="25.0"/>
    <param name="progress_window" value="6.0"/>
    <param name="min_progress" value="0.30"/>
    <param name="recent_goal_block_s" value="12.0"/>
    <param name="recent_goal_radius_m" value="0.6"/>
  </node>

</launch>
